hbase-server/src/test/java/org/apache/hadoop/hbase/io/hfile/TestCacheOnWrite.java

```diff
@@ -448,16 +455,34 @@ private void testNotCachingDataBlocksDuringCompactionInternals(boolean useTags)
       }
       region.flush(true);
     }
+
     clearBlockCache(blockCache);
     assertEquals(0, blockCache.getBlockCount());
+
     region.compact(false);
     LOG.debug("compactStores() returned");
 
+    boolean dataBlockCached = false;
     for (CachedBlock block: blockCache) {
-      assertNotEquals(BlockType.ENCODED_DATA, block.getBlockType());
-      assertNotEquals(BlockType.DATA, block.getBlockType());
+      if (BlockType.ENCODED_DATA.equals(block.getBlockType())
+          || BlockType.DATA.equals(block.getBlockType())) {
+        dataBlockCached = true;
+        break;
+      }
     }
+
+    // Data blocks should be cached in instances where we are caching blocks on write. In the case
+    // of testing
+    // BucketCache, we cannot verify block type as it is not stored in the cache.
+    assertTrue(
+      "\nTest description: " + testDescription + "\nprefetchCompactedBlocksOnWrite: "
+          + cacheBlocksOnCompaction + "\n",
+      (cacheBlocksOnCompaction && !(blockCache instanceof BucketCache)) == dataBlockCached);
+
     region.close();
+
+    // Clear the cache on write setting
+    conf.setBoolean(CacheConfig.CACHE_COMPACTED_BLOCKS_ON_WRITE_KEY, false);
```

Comment:
Minor : Get the existing conf value into a local var before setting and reset to that value here.